### PR TITLE
[STRAT-114] Route Model Refactor- Elevations Finalized 

### PR DESCRIPTION
**Pull request by krishna-kalavadia**
Elevations file completed. Also regenerated the Heartland Elevations because the last one had errors so that is also included in this PR, so only two files.

### CODE REVIEW

File: routemodel/data_retrieval/elevations.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    """
+    compressed_points = ""
+    lat = 0
+    long = 0
+    # Step 1: Start with a set of latitude and longitude values.
+    for index, pair in enumerate(coordinates):
+        for key in pair.keys():
+            # Step 2: Multiply each value by 100000 and round each result to the nearest int
+            new_lat = int(round(float(key) * 100000))
+            new_long = int(round(float(pair[key]) * 100000))
```
Comment (e-wai) [inline routemodel/data_retrieval/elevations.py:37]: Could swap `pair[key]` with `value` if you change line 34 to `for key, value in pair.items()`

File: routemodel/data_retrieval/elevations.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+            # Step 3: Calculate the difference between every pair of values
+            dx = (new_lat - lat)
```
Comment (e-wai) [inline routemodel/data_retrieval/elevations.py:40]: Sorry this is small, but don't think the brackets do anything for readability

File: routemodel/data_retrieval/elevations.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+            # Step 6: For each pair of latitude and longitude coordinates compute
+            #         ((latitude + longitude) * (latitude + longitude + 1) / 2) + latitude
+            index = int(((dx + dy) * (dx + dy + 1) / 2) + dx)
+
+            # Step 7: For each number, form a list of numbers by dividing the number by 32 repeatedly and recording
+            #         each remainder, stop when the quotient reaches zero
+            rem = []
+            while index > 0:
+                rem.append(int(index % 32))
```
Comment (e-wai) [inline routemodel/data_retrieval/elevations.py:63]: if you start with an int, think the int cast call is redundant

File: routemodel/data_retrieval/elevations.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+                index = int(index / 32)
```
Comment (e-wai) [inline routemodel/data_retrieval/elevations.py:64]: Can replace with `index // 32`

File: routemodel/data_retrieval/elevations.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+            # Step 8: Add 32 to each number except for the last number in each list.
+            for i in range(0, len(rem) - 1):
```
Comment (e-wai) [inline routemodel/data_retrieval/elevations.py:67]: python thing ->  can use something like

```suggestion
rem = [i + 32 for i in rem]
```

Comment (e-wai) [reply routemodel/data_retrieval/elevations.py:67]: Sorry I don't know how to use tabs in comments

File: routemodel/data_retrieval/elevations.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        url = BASE_URL + 'Elevation/List?'
+        # add parameters to URL
+        url += 'points={}&heights={}&key={}'.format(coordinates_str, heights, API_KEY)
+    elif method == 'polyline':
+        if sample_val <= 0:
+            print("Error, sample_val must be greater than 0")
+            sys.exit()
+        # append specified elevation URL details
+        url = BASE_URL + 'Elevation/Polyline?'
+        sample = str(sample_val)
```
Comment (e-wai) [inline routemodel/data_retrieval/elevations.py:101]: don't think you need the string cast; can just put it directly into format?

File: routemodel/data_retrieval/elevations.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+    if method == 'default':
+        # create DataFrame with headers
+        headers = ['Latitude', 'Longitude', 'Elevation']
+        elevations_df = pd.DataFrame(columns=headers)
+
+        # loop through coordinates and write coordinates into DataFrame
+        counter = 0
+        for index, pair in enumerate(coordinates):
+            for key in pair.keys():
```
Comment (e-wai) [inline routemodel/data_retrieval/elevations.py:146]: Same as first comment

**Review by e-wai — DISMISSED** [review]
Looks good to me. Can I assume that tests and `speedlimit` are a later issue?

**Comment by ufshk** [comment]
> 
> 
> Looks good to me. Can I assume that tests and `speedlimit` are a later issue?

Yes, yes you can.

File: routemodel/data_retrieval/elevations.py (hunk 5 above)
Comment (krishna-kalavadia) [reply routemodel/data_retrieval/elevations.py]: Resolved this over slack

**Review by e-wai — APPROVED** [review]
Looks really good to me 🥳 Really appreciate how you built everything

@ufshk have you reviewed too? do you have any thoughts on separating the `get` request into its own method for testing?

**Review by mdshiozaki — APPROVED** [review]
Looks solid to me, will wait to see about that comment on potentially moving the get request for testing but thank you!

**Comment by krishna-kalavadia** [comment]
> Looks really good to me 🥳 Really appreciate how you built everything
> 
> @ufshk have you reviewed too? do you have any thoughts on separating the `get` request into its own method for testing?



> Looks really good to me 🥳 Really appreciate how you built everything
> 
> @ufshk have you reviewed too? do you have any thoughts on separating the `get` request into its own method for testing?



> Looks solid to me, will wait to see about that comment on potentially moving the get request for testing but thank you!

Thanks for taking the time to review this!

**Pull request merged**